### PR TITLE
SMB version of MainActivityHelper.checkFolder()

### LINKED ISSUE #2189: Crash stacktrace at SMB file copy
Getting following stacktrace at log level if I try to copy file in SMB

```
2021-01-05 10:51:44.390 26559-26559/com.amaze.filemanager.debug W/System.err: java.io.FileNotFoundException: smb:/workgroup%3Bvishalnehra:Vishal%40123@192.168.0.126/vishalnehra/Downloads/DummyFile: open failed: ENOENT (No such file or directory)
2021-01-05 10:51:44.393 26559-26559/com.amaze.filemanager.debug W/System.err:     at libcore.io.IoBridge.open(IoBridge.java:496)
2021-01-05 10:51:44.393 26559-26559/com.amaze.filemanager.debug W/System.err:     at java.io.FileOutputStream.<init>(FileOutputStream.java:235)
2021-01-05 10:51:44.393 26559-26559/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.filesystem.FileUtil.isWritable(FileUtil.java:754)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.utils.MainActivityHelper.checkFolder(MainActivityHelper.java:420)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.finishCopying(PrepareCopyTask.java:325)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onEndDialog(PrepareCopyTask.java:275)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onEndDialog(PrepareCopyTask.java:269)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onPostExecute(PrepareCopyTask.java:186)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onPostExecute(PrepareCopyTask.java:64)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask.finish(AsyncTask.java:755)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask.access$900(AsyncTask.java:192)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:772)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:107)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at android.os.Looper.loop(Looper.java:214)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:7643)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
2021-01-05 10:51:44.394 26559-26559/com.amaze.filemanager.debug W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
2021-01-05 10:51:44.399 26559-26559/com.amaze.filemanager.debug W/System.err: Caused by: android.system.ErrnoException: open failed: ENOENT (No such file or directory)
```

### PR DESCRIPTION
**Pull request by TranceLove**
#### Issue tracker   
Fixes will automatically close the related issue

Fixes #2189

#### Release  
Addresses release/3.6
  
#### Test cases
- [x] Covered
  
#### Manual testing
- [x] Done  
  
If yes,  
- Device: Fairphone 3
- OS: LineageOS 16.0 (9.0)

#### Build tasks success  
Successfully running following tasks on local 
- [x] `./gradlew assembledebug`
- [x] `./gradlew spotlessCheck`

#### Additional Info
When copying from local to SMB, SMB URL was seen passed into MainActivityHelper.checkFolder(), causing the exception. Created a SMB version of the method to fix this.

### CODE REVIEW

File: app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/PrepareCopyTask.java [hunk 1]
```diff
@@ -322,7 +323,9 @@ private void finishCopying(
     }
 
     if (filesToCopyPerFolder.size() != 0) {
-      int mode = mainActivity.mainActivityHelper.checkFolder(new File(path), context);
+      int mode = -1;
+      if (OpenMode.SMB.equals(openMode)) mode = SmbUtil.checkFolder(path);
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/PrepareCopyTask.java:327]: Remember to put the body on another line and braces.

**Review by EmmanuelMess — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/PrepareCopyTask.java (hunk 1 above)
Comment (EmmanuelMess) [reply app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/PrepareCopyTask.java:327]: Otherwise accepted

**Review by EmmanuelMess — APPROVED** [review]

File: app/src/main/java/com/amaze/filemanager/utils/SmbUtil.java [hunk 2]
```diff
@@ -97,4 +108,21 @@ public static SmbFile create(String path) throws MalformedURLException {
...
+   * @param path SMB path
+   * @return {@link MainActivityHelper#DOESNT_EXIST} if specified SMB path doesn't exist on server,
+   *     else -1
+   */
+  public static int checkFolder(@NonNull String path) {
+    try {
+      SmbFile smbFile = create(path);
+      if (!smbFile.exists() || !smbFile.isDirectory()) return MainActivityHelper.DOESNT_EXIST;
+    } catch (SmbException | MalformedURLException e) {
+      Log.e(TAG, "Error checking folder existence", e);
```
Comment (VishalNehra) [inline app/src/main/java/com/amaze/filemanager/utils/SmbUtil.java:124]: Shouldn't we return `MainActivityHelper.DOESNT_EXIST` here as well?

File: app/src/main/java/com/amaze/filemanager/utils/SmbUtil.java [hunk 3]
```diff
@@ -97,4 +108,21 @@ public static SmbFile create(String path) throws MalformedURLException {
...
+    }
+    return -1;
```
Comment (VishalNehra) [inline app/src/main/java/com/amaze/filemanager/utils/SmbUtil.java:126]: Why return -1? Why not `MainActivityHelper.CAN_CREATE_FILES`?

File: app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/PrepareCopyTask.java [hunk 4]
```diff
@@ -322,7 +327,12 @@ private void finishCopying(
     }
 
     if (filesToCopyPerFolder.size() != 0) {
-      int mode = mainActivity.mainActivityHelper.checkFolder(new File(path), context);
+      int mode = -1;
+      if (OpenMode.SMB.equals(openMode)) {
```
Comment (VishalNehra) [inline app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/PrepareCopyTask.java:331]: The `checkFolder` is getting called from multiple places? This condition seem to be temporarily fixing only one of the places. My suggestion to use some form of abstraction or atleast make `checkFolder` generic to accept all paths and put this condition there.

**Review by VishalNehra — CHANGES_REQUESTED** [review]

**Comment by VishalNehra** [comment]
I can't seem to figure out why the reported crash has path as `java.io.FileNotFoundException: smb:/workgroup` with a single forward slash 🤔

File: app/src/main/java/com/amaze/filemanager/utils/SmbUtil.java (hunk 3 above)
Comment (TranceLove) [reply app/src/main/java/com/amaze/filemanager/utils/SmbUtil.java]: Good choice, but it's not over yet. When test on Fairphone 3 running LineageOS 16.0 (9.0) when returning `CAN_CREATE_FILES` files will not be overwritten - no actions at all; when return -1 seems will be overwritten without warning or dialogs.

Seems to be safest answer, but it's just not over yet. :/

**Comment by TranceLove** [comment]
> I can't seem to figure out why the reported crash has path as java.io.FileNotFoundException: smb:/workgroup with a single forward slash

PrepareCopyTask called this, that the SMB URI passed into the `checkFolder` method. Yeap, don't know why...

**Comment by VishalNehra** [comment]
## Issue explanation (write below this line)
Trying to copy file to smb with the above changeset


## Exception
* __User Action:__ UI Error
* __Request:__ Application crash
* __Version:__ 3.6.0
* __OS:__ Linux Android 10 - 29
* __Device:__ OnePlus6T
* __Model:__ ONEPLUS A6010
* __Product:__ OnePlus6T
<details><summary><b>Crash log </b></summary><p>

```
android.os.NetworkOnMainThreadException
	at android.os.StrictMode$AndroidBlockGuardPolicy.onNetwork(StrictMode.java:1565)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:116)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:161)
	at jcifs.smb.SmbTransportImpl.doSend(SmbTransportImpl.java:910)
	at jcifs.util.transport.Transport.doSend(Transport.java:276)
	at jcifs.util.transport.Transport.sendrecv(Transport.java:215)
	at jcifs.smb.SmbTransportImpl.sendrecv(SmbTransportImpl.java:1021)
	at jcifs.smb.SmbTransportImpl.send(SmbTransportImpl.java:1539)
	at jcifs.smb.SmbSessionImpl.send(SmbSessionImpl.java:409)
	at jcifs.smb.SmbTreeImpl.send(SmbTreeImpl.java:472)
	at jcifs.smb.SmbTreeConnection.send0(SmbTreeConnection.java:404)
	at jcifs.smb.SmbTreeConnection.send(SmbTreeConnection.java:318)
	at jcifs.smb.SmbTreeConnection.send(SmbTreeConnection.java:298)
	at jcifs.smb.SmbTreeHandleImpl.send(SmbTreeHandleImpl.java:130)
	at jcifs.smb.SmbTreeHandleImpl.send(SmbTreeHandleImpl.java:117)
	at jcifs.smb.SmbFile.withOpen(SmbFile.java:1741)
	at jcifs.smb.SmbFile.withOpen(SmbFile.java:1710)
	at jcifs.smb.SmbFile.withOpen(SmbFile.java:1704)
	at jcifs.smb.SmbFile.queryPath(SmbFile.java:770)
	at jcifs.smb.SmbFile.exists(SmbFile.java:851)
	at com.amaze.filemanager.utils.SmbUtil.checkFolder(SmbUtil.java:122)
	at com.amaze.filemanager.utils.MainActivityHelper.checkFolder(MainActivityHelper.java:413)
	at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.finishCopying(PrepareCopyTask.java:329)
	at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onEndDialog(PrepareCopyTask.java:279)
	at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onEndDialog(PrepareCopyTask.java:272)
	at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onPostExecute(PrepareCopyTask.java:186)
	at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onPostExecute(PrepareCopyTask.java:64)
	at android.os.AsyncTask.finish(AsyncTask.java:755)
	at android.os.AsyncTask.access$900(AsyncTask.java:192)
	at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:772)
	at android.os.Handler.dispatchMessage(Handler.java:107)
	at android.os.Looper.loop(Looper.java:214)
	at android.app.ActivityThread.main(ActivityThread.java:7643)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)

```
</details>
<hr>

File: app/src/main/java/com/amaze/filemanager/filesystem/FolderState.kt [hunk 5]
```diff
@@ -27,6 +27,7 @@ const val WRITABLE_OR_ON_SDCARD = 1
 
 // For Android 5
 const val CAN_CREATE_FILES = 2
+const val WRITABLE_ON_REMOTE = 3
```
Comment (VishalNehra) [inline app/src/main/java/com/amaze/filemanager/filesystem/FolderState.kt:30]: Is introducing new const okay? I see many operations like delete Files, compress, extract, that are calling `checkFolder` displays `Operation not supported` in their else cases.

**Review by VishalNehra — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/amaze/filemanager/filesystem/FolderState.kt (hunk 5 above)
Comment (TranceLove) [reply app/src/main/java/com/amaze/filemanager/filesystem/FolderState.kt:30]: Flag was introduced so https://github.com/TranceLove/AmazeFileManager/blob/ea5aca20d0151b421a687c21944615cffb475d5b/app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/PrepareCopyTask.java#L331 would pass.

MainActivityHelper.deleteFiles() automatically goes to DeleteTask right away if the file(s) is/are SMB. See https://github.com/TranceLove/AmazeFileManager/blob/ea5aca20d0151b421a687c21944615cffb475d5b/app/src/main/java/com/amaze/filemanager/utils/MainActivityHelper.java#L630-L633

But for extract/compress files, saw the problem - can work more to accommodate the flag.

**Comment by VishalNehra** [comment]
@TranceLove can you please fix the conflict here

**Comment by VishalNehra** [comment]
Sadly I'm still getting the exception log, so doesn't fix the issue still.

```
2021-02-14 02:16:19.112 31870-31870/com.amaze.filemanager.debug W/System.err: java.io.FileNotFoundException: /smb:/vishalnehra:Vishal%40123@192.168.3.102/vishalnehra/Downloads/2020-07-02_14.07 (1)/DummyFile: open failed: ENOENT (No such file or directory)
2021-02-14 02:16:19.115 31870-31870/com.amaze.filemanager.debug W/System.err:     at libcore.io.IoBridge.open(IoBridge.java:496)
2021-02-14 02:16:19.115 31870-31870/com.amaze.filemanager.debug W/System.err:     at java.io.FileOutputStream.<init>(FileOutputStream.java:235)
2021-02-14 02:16:19.115 31870-31870/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.filesystem.FileUtil.isWritable(FileUtil.java:754)
2021-02-14 02:16:19.115 31870-31870/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.utils.MainActivityHelper.checkFolder(MainActivityHelper.java:435)
2021-02-14 02:16:19.115 31870-31870/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.utils.MainActivityHelper.checkFolder(MainActivityHelper.java:414)
2021-02-14 02:16:19.115 31870-31870/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.finishCopying(PrepareCopyTask.java:331)
2021-02-14 02:16:19.115 31870-31870/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onEndDialog(PrepareCopyTask.java:281)
2021-02-14 02:16:19.115 31870-31870/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onEndDialog(PrepareCopyTask.java:274)
2021-02-14 02:16:19.115 31870-31870/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onPostExecute(PrepareCopyTask.java:188)
2021-02-14 02:16:19.115 31870-31870/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.asynctasks.PrepareCopyTask.onPostExecute(PrepareCopyTask.java:66)
2021-02-14 02:16:19.116 31870-31870/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask.finish(AsyncTask.java:755)
2021-02-14 02:16:19.116 31870-31870/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask.access$900(AsyncTask.java:192)
2021-02-14 02:16:19.116 31870-31870/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:772)
2021-02-14 02:16:19.116 31870-31870/com.amaze.filemanager.debug W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:107)
2021-02-14 02:16:19.116 31870-31870/com.amaze.filemanager.debug W/System.err:     at android.os.Looper.loop(Looper.java:214)
2021-02-14 02:16:19.116 31870-31870/com.amaze.filemanager.debug W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:7643)
2021-02-14 02:16:19.116 31870-31870/com.amaze.filemanager.debug W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
2021-02-14 02:16:19.116 31870-31870/com.amaze.filemanager.debug W/System.err:     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
2021-02-14 02:16:19.116 31870-31870/com.amaze.filemanager.debug W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
2021-02-14 02:16:19.119 31870-31870/com.amaze.filemanager.debug W/System.err: Caused by: android.system.ErrnoException: open failed: ENOENT (No such file or directory)
2021-02-14 02:16:19.119 31870-31870/com.amaze.filemanager.debug W/System.err:     at libcore.io.Linux.open(Native Method)
2021-02-14 02:16:19.121 31870-31870/com.amaze.filemanager.debug W/System.err:     at libcore.io.ForwardingOs.open(ForwardingOs.java:167)
2021-02-14 02:16:19.121 31870-31870/com.amaze.filemanager.debug W/System.err:     at libcore.io.BlockGuardOs.open(BlockGuardOs.java:252)
2021-02-14 02:16:19.121 31870-31870/com.amaze.filemanager.debug W/System.err:     at libcore.io.ForwardingOs.open(ForwardingOs.java:167)
2021-02-14 02:16:19.121 31870-31870/com.amaze.filemanager.debug W/System.err:     at android.app.ActivityThread$AndroidOs.open(ActivityThread.java:7542)
2021-02-14 02:16:19.121 31870-31870/com.amaze.filemanager.debug W/System.err:     at libcore.io.IoBridge.open(IoBridge.java:482)
2021-02-14 02:16:19.121 31870-31870/com.amaze.filemanager.debug W/System.err: 	... 18 more
```

**Comment by EmmanuelMess** [comment]
@TranceLove this needs to be rebased

**Comment by EmmanuelMess** [comment]
This seems to have been rebased before the build fix.

**Review by EmmanuelMess — APPROVED** [review]
Fixes issue and code seems ok

**Review by VishalNehra — APPROVED** [review]

**Pull request merged**